### PR TITLE
[Chore] Fix the issue that the click version is locked at 7.x

### LINKED ISSUE #123: Update Click to 8.x
**Describe the feature**

Can we upgrade click to 8.x?

**Motivation**

Click 7.x is somehow old and other packages (such as black) migrates to use click 8.x, when install both there will be a conflict. So can we make a upgrade?

### PR DESCRIPTION
**Pull request by ice-tong**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation
Fix the issue that the click version is locked at 7.x, fixes #123
In addition, add .vscode into .gitignore make vscode user happy.

## Modification
- .gitignore
- setup.py
- requirements/install.txt
- mim/click/\_\_init\_\_.py
- mim/click/compat.py
- mim/click/option.py
- mim/commands/download.py
- mim/commands/install.py
- mim/commands/search.py
- mim/commands/uninstall.py

## BC-breaking (Optional)
No

## Use cases (Optional)
No

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
3. If the modification has potential influence on downstream projects, this PR should be tested with downstream projects, like MMDet or MMCls.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

File: mim/click/compat.py [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+    return shell_complete
+
+
+def argument(*param_decls, **attrs):
+    """A decorator compatible with click 7.x and 8.x.
+
+    Same as ``click.argument``.
+    """
+    if is_click_ge_8():
```
Comment (zhouzaida) [inline mim/click/compat.py:48]:
```suggestion
    if LooseVersion(click.__version__) >= LooseVersion('8.0.0'):
```

File: mim/click/compat.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+from distutils.version import LooseVersion
+
+import click
+
+
+def is_click_ge_8():
+    """Check if the click version is greater than or equal to 8.x."""
+    return LooseVersion(click.__version__) >= LooseVersion('8.0.0')
```
Comment (zhouzaida) [inline mim/click/compat.py:9]: This helper function can be removed.

Comment (ice-tong) [reply mim/click/compat.py]: Done!

File: mim/commands/download.py [hunk 3]
```diff
@@ -4,7 +4,9 @@
 
 import click
 
-from mim.click import OptionEatAll, get_downstream_package, param2lowercase
+from mim.click import OptionEatAll
+from mim.click import argument as compat_argument
```
Comment (zhouzaida) [inline mim/commands/download.py:8]: We can directly use argument rather than compat_argument.

File: mim/commands/download.py [hunk 4]
```diff
@@ -4,7 +4,9 @@
...
+from mim.click import get_downstream_package, param2lowercase
```
Comment (zhouzaida) [inline mim/commands/download.py:9]:
```suggestion
from mim.click import (OptionEatAll, argument, get_downstream_package,
                       param2lowercase)
```

**Comment by zhouzaida** [comment]
#128 has been merged and please update your branch to fix the CI.

In addition, it would be better to test `click < 8.0` in CI.

**Comment by ice-tong** [comment]
> Suggest renaming `mim/click/compat.py` to `mim/click/compatibility .py`

The name `compat.py` are much more commonly used than `compatibility.py`.
For examples:
- [requests/compat.py](https://github.com/psf/requests/blob/main/requests/compat.py)
- [pytest/compat.py](https://github.com/pytest-dev/pytest/blob/main/src/_pytest/compat.py)
- [tensorflow/compat/compat.py](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/compat/compat.py)

File: mim/commands/download.py (hunk 4 above)
Comment (ice-tong) [reply mim/commands/download.py]: Done

File: mim/commands/download.py (hunk 3 above)
Comment (ice-tong) [reply mim/commands/download.py]: Done

**Comment by ice-tong** [comment]
> #128 has been merged and please update your branch to fix the CI.
> 
> In addition, it would be better to test `click < 8.0` in [CI](https://github.com/open-mmlab/mim/blob/main/.github/workflows/build.yml).
> 
> ```yaml
>       - name: Test click==7.1.2
>         run: python -m pip install click==7.1.2
>         if: ${{matrix.python-version == '3.8'}}
> ```

Add a step in CI： `Run unittests with click < 8.0.0`
```
- name: Run unittests with click < 8.0.0
   run: |
       python -m pip install click==7.1.2
       pytest tests/
```

File: mim/click/option.py [hunk 5]
```diff
@@ -45,3 +45,14 @@ def parser_process(value, state):
                 our_parser.process = parser_process
                 break
         return retval
+
+    def type_cast_value(self, ctx, value):
+        """Override the `type_cast_value` method.
+
```
Comment (zhouzaida) [inline mim/click/option.py:51]:
```suggestion
        """Convert and validate a value against the option's
        :attr:`type`, :attr:`multiple`, and :attr:`nargs`.

```

File: mim/click/compat.py [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+
+    return shell_complete
+
+
+def argument(*param_decls, **attrs):
+    """A decorator compatible with click 7.x and 8.x.
+
+    Same as ``click.argument``.
+    """
+    if LooseVersion(click.__version__) >= LooseVersion('8.0.0'):
```
Comment (zhouzaida) [inline mim/click/compat.py:43]:
```suggestion
    # 'autocompletion' will be removed in Click 8.1 and its new name is
    # 'shell_complete'.
    if LooseVersion(click.__version__) >= LooseVersion('8.0.0'):
```

File: mim/click/compat.py [hunk 7]
```diff
@@ -0,0 +1,49 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+from distutils.version import LooseVersion
+
+import click
+
+
+def autocompletion_to_shell_complete(autocompletion):
```
Comment (zhouzaida) [inline mim/click/compat.py:7]: It seems like the autocompletion parameter is not used.

File: mim/click/compat.py [hunk 8]
```diff
@@ -0,0 +1,49 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+from distutils.version import LooseVersion
+
```
Comment (zhouzaida) [inline mim/click/compat.py:3]: from typing import Callable

File: mim/click/compat.py (hunk 7 above)
Comment (zhouzaida) [inline mim/click/compat.py:7]:
```suggestion
def autocompletion_to_shell_complete(autocompletion: Callable):
```

File: mim/click/compat.py [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+
+def autocompletion_to_shell_complete(autocompletion):
+    """Convert autocompletion to shell_complete.
+
+    Reference:
+    https://github.com/pallets/click/blob/8.0.0/src/click/core.py#L2059
+
+    Args:
+        autocompletion: A function that returns custom shell completions.
```
Comment (zhouzaida) [inline mim/click/compat.py:14]:
```suggestion
        autocompletion (callable): A function that returns custom shell completions.
```

**Pull request merged**